pkg/auth/providers/azuread/azuread_client.go

```diff
@@ -0,0 +1,550 @@
+	if err != nil {
+		logrus.Errorf("AzureAD getAccessToken: GET url %v received error from azure, err: %v", url, err)
+		return []Account{}, nil, err
+	}
+	if newAccessToken != "" {
+		resp, _, providerInfo, err = ac.getFromAzureAD(newAccessToken, config, url)
+		if err != nil {
+			logrus.Errorf("AzureAD getAccessToken: GET url %v received error from azure, err: %v", url, err)
+			return []Account{}, nil, err
+		}
+	}
+	defer resp.Body.Close()
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	result := &searchResult{}
+	if err := json.Unmarshal(b, result); err != nil {
+		return nil, nil, err
+	}
+	return result.Items, providerInfo, nil
+}
+
+func (ac *Client) getUserByID(id string, azureAccessToken string, config *v3.AzureADConfig) (Account, map[string]string, error) {
+	if azureAccessToken == "" {
+		return Account{}, nil, httperror.NewAPIError(httperror.ServerError, "No Azure Access token")
+	}
+	authority, graphEndpoint, _, err := ac.uriEndpointConstruction(config)
+	if err != nil {
+		logrus.Errorf("Wrong Azure Domain %v is provided", config.Domain)
+		return Account{}, nil, fmt.Errorf("Wrong Azure Domain %v is provided", config.Domain)
+	}
+
+	url, err := ac.getURL(authority, graphEndpoint, "USERS", config, id)
+
+	resp, newAccessToken, providerInfo, err := ac.getFromAzureAD(azureAccessToken, config, url)
+	if err != nil {
+		logrus.Errorf("AzureAD getAzureADUser: GET url %v received error from Azure, err: %v", url, err)
+		return Account{}, nil, err
+	}
+	if newAccessToken != "" {
+		resp, _, providerInfo, err = ac.getFromAzureAD(newAccessToken, config, url)
+		if err != nil {
+			logrus.Errorf("AzureAD getAccessToken: GET url %v received error from azure, err: %v", url, err)
+			return Account{}, nil, err
+		}
+	}
+	defer resp.Body.Close()
+	var azureADAcct Account
+
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		logrus.Errorf("AzureAD getAzureADUser: error reading response, err: %v", err)
+		return Account{}, nil, err
+	}
+
+	if err := json.Unmarshal(b, &azureADAcct); err != nil {
+		logrus.Errorf("AzureAD getAzureADUser: error unmarshalling response, err: %v", err)
+		return Account{}, nil, err
+	}
+
+	return azureADAcct, providerInfo, nil
+}
+
+func (ac *Client) getGroupByID(id string, azureAccessToken string, config *v3.AzureADConfig) (Account, map[string]string, error) {
+	if azureAccessToken == "" {
+		return Account{}, nil, httperror.NewAPIError(httperror.ServerError, "No Azure Access token")
+	}
+	authority, graphEndpoint, _, err := ac.uriEndpointConstruction(config)
+	if err != nil {
+		logrus.Errorf("Wrong Azure Domain %v is provided", config.Domain)
+		return Account{}, nil, fmt.Errorf("Wrong Azure Domain %v is provided", config.Domain)
+	}
+
+	url, err := ac.getURL(authority, graphEndpoint, "GROUPS", config, id)
+
+	resp, newAccessToken, providerInfo, err := ac.getFromAzureAD(azureAccessToken, config, url)
```

Comment:
not closed properly. you re-assign resp at line 395, but you didnt close the body on this instance of it. you need to do a close immediately after the err check